Some primary key daily test failed

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
nosetests-3.4 -s --with-xunit --xunit-file=./result/TestUpdateBasic_test_delete_using_expr.xml --verbose case/system_case/test_dml/test_update/test_update_basic.py:TestUpdateBasic.test_delete_using_expr --nologcapture

### Expected behavior (Required)
Test OK
### Real behavior (Required)
```
[jenkins@ci01 StarRocksTest]$ nosetests-3.4 -s --with-xunit --xunit-file=./result/TestUpdateBasic_test_delete_using_expr.xml --verbose case/system_case/test_dml/test_update/test_update_basic.py:TestUpdateBasic.test_delete_using_expr --nologcapture
test_update_basic.TestUpdateBasic.test_delete_using_expr ... select connection_id()
set global is_report_success = true
create database update_test_db_fc28b754_ac11_11ec_9ed2_00163e0e489a
use update_test_db_fc28b754_ac11_11ec_9ed2_00163e0e489a
create table delete_using_expr (`pk` int not null,`col2` int not null) primary key (pk) DISTRIBUTED BY HASH(pk) BUCKETS 1 PROPERTIES("replication_num" = "1");
===================       srccol0,srccol1,srccol2,pk=srccol0,col2=srccol1,__op=srccol2
curl --location-trusted -u root: -T /tmp/tmpj6xwgr7z -XPUT -H label:stream_load_m2_1648195378755 -H column_separator:, -H columns:srccol0,srccol1,srccol2,pk=srccol0,col2=srccol1,__op=srccol2 http://172.26.92.141:8034/api/update_test_db_fc28b754_ac11_11ec_9ed2_00163e0e489a/delete_using_expr/_stream_load
** update/delete data orig:0, update:1000 delete:142 now:870
select count(1), sum(col2) from delete_using_expr
drop database update_test_db_fc28b754_ac11_11ec_9ed2_00163e0e489a
FAIL

======================================================================
FAIL: test_update_basic.TestUpdateBasic.test_delete_using_expr
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/disk1/jenkins/workspace/master_release/StarRocksTest/case/system_case/test_dml/test_update/test_update_basic.py", line 725, in test_delete_using_expr
    eq_(result["result"][0], (len(current_data[0]), none_sum(current_data[col2col])))
AssertionError: (946, 31037958) != (870, 28551167)

----------------------------------------------------------------------
XML: /home/disk1/jenkins/workspace/master_release/StarRocksTest/result/TestUpdateBasic_test_delete_using_expr.xml
----------------------------------------------------------------------
Ran 1 test in 0.515s

```
### StarRocks version (Required)
current main

--- a/be/src/exec/vectorized/sorting/sort_column.cpp
+++ b/be/src/exec/vectorized/sorting/sort_column.cpp
@@ -8,6 +8,7 @@
 #include "exec/vectorized/sorting/sort_helper.h"
 #include "exec/vectorized/sorting/sort_permute.h"
 #include "exec/vectorized/sorting/sorting.h"
+#include "util/orlp/pdqsort.h"
 
 namespace starrocks::vectorized {
 
@@ -127,8 +128,8 @@ Status sort_and_tie_columns(const bool& cancel, const Columns& columns, const st
     return Status::OK();
 }
 
-Status sort_and_tie_columns(const bool& cancel, const Columns& columns, const std::vector<int>& sort_orders,
-                            const std::vector<int>& null_firsts, SmallPermutation* small_perm) {
+Status stable_sort_and_tie_columns(const bool& cancel, const Columns& columns, const std::vector<int>& sort_orders,
+                                   const std::vector<int>& null_firsts, SmallPermutation* small_perm) {
     if (columns.size() < 1) {
         return Status::OK();
     }
@@ -144,6 +145,17 @@ Status sort_and_tie_columns(const bool& cancel, const Columns& columns, const st
         bool build_tie = col_index != columns.size() - 1;
         RETURN_IF_ERROR(
                 sort_and_tie_column(cancel, column, is_asc_order, is_null_first, *small_perm, tie, range, build_tie));
+    }
+
+    // Stable sort need extra runs of sorting on permutation
+    TieIterator ti(tie);
+    while (ti.next()) {
+        int range_first = ti.range_first, range_last = ti.range_last;
+        if (range_last - range_first > 1) {
+            ::pdqsort(
+                    cancel, small_perm->begin() + range_first, small_perm->begin() + range_last,
+                    [](SmallPermuteItem lhs, SmallPermuteItem rhs) { return lhs.index_in_chunk < rhs.index_in_chunk; });
+        }
     }
 
     return Status::OK();

--- a/be/src/exec/vectorized/sorting/sort_column.cpp
+++ b/be/src/exec/vectorized/sorting/sort_column.cpp
@@ -142,9 +142,8 @@ Status stable_sort_and_tie_columns(const bool& cancel, const Columns& columns, c
         ColumnPtr column = columns[col_index];
         bool is_asc_order = (sort_orders[col_index] == 1);
         bool is_null_first = is_asc_order ? (null_firsts[col_index] == -1) : (null_firsts[col_index] == 1);
-        bool build_tie = col_index != columns.size() - 1;
         RETURN_IF_ERROR(
-                sort_and_tie_column(cancel, column, is_asc_order, is_null_first, *small_perm, tie, range, build_tie));
+                sort_and_tie_column(cancel, column, is_asc_order, is_null_first, *small_perm, tie, range, true));
     }
 
     // Stable sort need extra runs of sorting on permutation

--- a/be/src/exec/vectorized/sorting/sorting.h
+++ b/be/src/exec/vectorized/sorting/sorting.h
@@ -20,8 +20,10 @@ Status sort_and_tie_column(const bool& cancel, const ColumnPtr column, bool is_a
 // Sort multiple columns using column-wise algorithm, output the order in permutation array
 Status sort_and_tie_columns(const bool& cancel, const Columns& columns, const std::vector<int>& sort_orders,
                             const std::vector<int>& null_firsts, Permutation* permutation);
-Status sort_and_tie_columns(const bool& cancel, const Columns& columns, const std::vector<int>& sort_orders,
-                            const std::vector<int>& null_firsts, SmallPermutation* permutation);
+
+// Sort multiple columns, and stable
+Status stable_sort_and_tie_columns(const bool& cancel, const Columns& columns, const std::vector<int>& sort_orders,
+                                   const std::vector<int>& null_firsts, SmallPermutation* permutation);
 
 // Compare the column with the `rhs_value`, which must have the some type with column.
 // @param cmp_result compare result is written into this array, value must within -1,0,1

--- a/be/src/storage/vectorized/memtable.cpp
+++ b/be/src/storage/vectorized/memtable.cpp
@@ -327,7 +327,7 @@ void MemTable::_sort_column_inc() {
         null_firsts.push_back(-1);
     }
 
-    Status st = sort_and_tie_columns(false, columns, sort_orders, null_firsts, &_permutations);
+    Status st = stable_sort_and_tie_columns(false, columns, sort_orders, null_firsts, &_permutations);
     CHECK(st.ok());
 }
 

--- a/be/test/exec/vectorized/chunks_sorter_test.cpp
+++ b/be/test/exec/vectorized/chunks_sorter_test.cpp
@@ -543,17 +543,22 @@ TEST_F(ChunksSorterTest, stable_sort) {
     ColumnPtr col1 = ColumnHelper::create_column(type_desc, false);
     ColumnPtr col2 = ColumnHelper::create_column(type_desc, false);
     Columns columns{col1, col2};
-    std::vector<int32_t> elements{3, 1, 1, 2, 1, 2, 3};
-    for (auto x : elements) {
-        col1->append_datum(Datum(x));
-        col2->append_datum(Datum(x));
+    std::vector<int32_t> elements_col1{3, 1, 1, 2, 1, 2, 3};
+    std::vector<int32_t> elements_col2{3, 2, 1, 3, 1, 2, 3};
+    for (int i = 0; i < elements_col1.size(); i++) {
+        col1->append_datum(Datum(elements_col1[i]));
+        col2->append_datum(Datum(elements_col2[i]));
     }
 
     SmallPermutation perm = create_small_permutation(N);
-    stable_sort_and_tie_columns(false, columns, {1}, {1}, &perm);
+    stable_sort_and_tie_columns(false, columns, {1, 1}, {1, 1}, &perm);
 
     bool sorted = std::is_sorted(perm.begin(), perm.end(), [&](SmallPermuteItem lhs, SmallPermuteItem rhs) {
         int x = col1->compare_at(lhs.index_in_chunk, rhs.index_in_chunk, *col1, 1);
+        if (x != 0) {
+            return x < 0;
+        }
+        x = col2->compare_at(lhs.index_in_chunk, rhs.index_in_chunk, *col2, 1);
         if (x != 0) {
             return x < 0;
         }
@@ -562,7 +567,7 @@ TEST_F(ChunksSorterTest, stable_sort) {
     ASSERT_TRUE(sorted);
     std::vector<uint32_t> result;
     permutate_to_selective(perm, &result);
-    std::vector<uint32_t> expect{1, 2, 4, 3, 5, 0, 6};
+    std::vector<uint32_t> expect{2, 4, 1, 5, 3, 0, 6};
     ASSERT_EQ(expect, result);
 }
 

--- a/be/test/exec/vectorized/chunks_sorter_test.cpp
+++ b/be/test/exec/vectorized/chunks_sorter_test.cpp
@@ -8,6 +8,7 @@
 #include "exec/vectorized/chunks_sorter_topn.h"
 #include "exec/vectorized/sorting/sort_helper.h"
 #include "exec/vectorized/sorting/sort_permute.h"
+#include "exec/vectorized/sorting/sorting.h"
 #include "exprs/slot_ref.h"
 #include "fmt/core.h"
 #include "runtime/runtime_state.h"
@@ -534,6 +535,35 @@ static void reset_permutation(SmallPermutation& permutation, int n) {
     for (int i = 0; i < permutation.size(); i++) {
         permutation[i].index_in_chunk = i;
     }
+}
+
+TEST_F(ChunksSorterTest, stable_sort) {
+    constexpr int N = 7;
+    TypeDescriptor type_desc = TypeDescriptor(TYPE_INT);
+    ColumnPtr col1 = ColumnHelper::create_column(type_desc, false);
+    ColumnPtr col2 = ColumnHelper::create_column(type_desc, false);
+    Columns columns{col1, col2};
+    std::vector<int32_t> elements{3, 1, 1, 2, 1, 2, 3};
+    for (auto x : elements) {
+        col1->append_datum(Datum(x));
+        col2->append_datum(Datum(x));
+    }
+
+    SmallPermutation perm = create_small_permutation(N);
+    stable_sort_and_tie_columns(false, columns, {1}, {1}, &perm);
+
+    bool sorted = std::is_sorted(perm.begin(), perm.end(), [&](SmallPermuteItem lhs, SmallPermuteItem rhs) {
+        int x = col1->compare_at(lhs.index_in_chunk, rhs.index_in_chunk, *col1, 1);
+        if (x != 0) {
+            return x < 0;
+        }
+        return lhs.index_in_chunk < rhs.index_in_chunk;
+    });
+    ASSERT_TRUE(sorted);
+    std::vector<uint32_t> result;
+    permutate_to_selective(perm, &result);
+    std::vector<uint32_t> expect{1, 2, 4, 3, 5, 0, 6};
+    ASSERT_EQ(expect, result);
 }
 
 TEST_F(ChunksSorterTest, column_incremental_sort) {

